lib/location/method_gnss.c

```diff
@@ -370,93 +394,156 @@ static void method_gnss_pgps_request_work_fn(struct k_work *item)
 {
-	int type_count = 0;
+	int32_t time_since_agps_req;
 
-#if !defined(CONFIG_NRF_CLOUD_PGPS)
-	/* If P-GPS is enabled, use predicted ephemeris to save power instead of requesting them
-	 * using A-GPS.
-	 */
-	if (request->sv_mask_ephe) {
-		type_count++;
-	}
-	if (request->sv_mask_alm) {
-		type_count++;
-	}
-#endif
-	if (request->data_flags & NRF_MODEM_GNSS_AGPS_GPS_UTC_REQUEST) {
-		type_count++;
-	}
-	if (request->data_flags & NRF_MODEM_GNSS_AGPS_KLOBUCHAR_REQUEST) {
-		type_count++;
-	}
-	if (request->data_flags & NRF_MODEM_GNSS_AGPS_NEQUICK_REQUEST) {
-		type_count++;
-	}
-	if (request->data_flags & NRF_MODEM_GNSS_AGPS_SYS_TIME_AND_SV_TOW_REQUEST) {
-		type_count++;
-	}
-	if (request->data_flags & NRF_MODEM_GNSS_AGPS_POSITION_REQUEST) {
-		type_count++;
-	}
-	if (request->data_flags &  NRF_MODEM_GNSS_AGPS_INTEGRITY_REQUEST) {
-		type_count++;
-	}
-
-	if (type_count == 0) {
+	/* Check if A-GPS data is needed. */
+	if (request->sv_mask_ephe == 0 && request->sv_mask_alm == 0 && request->data_flags == 0) {
 		LOG_DBG("No A-GPS data types requested");
 		return false;
-	} else {
-		return true;
 	}
+
+	/* A-GPS data is needed, check if enough time has passed since the last A-GPS data
+	 * request.
+	 */
+	if (agps_req_timestamp != 0) {
+		time_since_agps_req = k_uptime_get() - agps_req_timestamp;
+		if (time_since_agps_req < (AGPS_REQUEST_MIN_INTERVAL * MSEC_PER_SEC)) {
+			LOG_DBG("Skipping A-GPS request, time since last request: %d s",
+				time_since_agps_req / 1000);
+			return false;
+		}
+	}
+
+	return true;
 }
 #endif
 
 #if defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS)
-static void method_gnss_request_assistance(void)
+/* Triggers A-GPS data request and/or injection of P-GPS predictions.
+ *
+ * Before this function is called, the assistance data need from GNSS must be stored into
+ * 'agps_request'.
+ *
+ * There are three possible assistance configurations with different behavior:
+ *
+ * A-GPS only:
+ * Ephemerides, almanacs and additional assistance data is handled by A-GPS.
+ * If any additional assistance data (UTC, Klobuchar, GPS time, integrity or position) is needed,
+ * all additional assistance is requested at the same time.
+ *
+ * A-GPS and P-GPS:
+ * Ephemerides are handled by P-GPS.
+ * Almanacs are not used.
+ * Additional assistance data is handled by A-GPS.
+ * If any additional assistance data (UTC, Klobuchar, GPS time, integrity or position) is needed,
+ * all additional assistance is requested at the same time.
+ *
+ * P-GPS only:
+ * Ephemerides are handled by P-GPS.
+ * Almanacs are not used.
+ * GPS time and position assistance are handled by P-GPS.
+ *
+ * The frequency of A-GPS data requests is limited to avoid requesting data again and again for
+ * example in case the server is down.
+ *
+ * The reason why almanacs are not used with P-GPS is, that GNSS always has valid (and more
+ * accurate) ephemerides available. Thus, there's no benefit from having almanacs.
```

Comment:
As discussed offline, I'll change the implementation so, that almanacs are downloaded in case predictions are downloaded for a period longer than one week. This way almanacs can be used at least to help satellite acquisition if predicted ephemerides are not available.